goon/code/datums/browserOutput.dm

```diff
@@ -0,0 +1,283 @@
+// (This relies on byond automatically storing icons in savefiles as base64)
+/proc/icon2base64(var/icon/icon, var/iconKey = "misc")
+	if (!isicon(icon))
+		return 0
+
+	iconCache[iconKey] << icon
+	var/iconData = iconCache.ExportText(iconKey)
+	var/list/partial = splittext(iconData, "{")
+	return replacetext(copytext(partial[2], 3, -5), "\n", "")
+
+/proc/bicon(var/obj)
+	if (!obj)
+		return
+
+	if (isicon(obj))
+		//Icons get pooled constantly, references are no good here.
+		/*if (!bicon_cache["\ref[obj]"]) // Doesn't exist yet, make it.
+			bicon_cache["\ref[obj]"] = icon2base64(obj)
+		return "<img class='icon misc' src='data:image/png;base64,[bicon_cache["\ref[obj]"]]'>"*/
+		return "<img class='icon misc' src='data:image/png;base64,[icon2base64(obj)]'>"
+
+	// Either an atom or somebody fucked up and is gonna get a runtime, which I'm fine with.
+	var/atom/A = obj
+	var/key = "[istype(A.icon, /icon) ? "\ref[A.icon]" : A.icon]:[A.icon_state]"
+	if (!bicon_cache[key]) // Doesn't exist, make it.
+		var/icon/I = icon(A.icon, A.icon_state, SOUTH, 1)
+		if (ishuman(obj)) // Shitty workaround for a BYOND issue.
+			var/icon/temp = I
+			I = icon()
+			I.Insert(temp, dir = SOUTH)
+		bicon_cache[key] = icon2base64(I, key)
+
+	return "<img class='icon [A.icon_state]' src='data:image/png;base64,[bicon_cache[key]]'>"
+
+//Aliases for bicon
+/proc/bi(obj)
+	bicon(obj)
+
+//Costlier version of bicon() that uses getFlatIcon() to account for overlays, underlays, etc. Use with extreme moderation, ESPECIALLY on mobs.
+/proc/costly_bicon(var/obj)
+	if (!obj)
+		return
+
+	if (isicon(obj))
+		return bicon(obj)
+
+	var/icon/I = getFlatIcon(obj)
+	return bicon(I)
+
+/proc/to_chat(target, message)
+	//Ok so I did my best but I accept that some calls to this will be for shit like sound and images
+	//It stands that we PROBABLY don't want to output those to the browser output so just handle them here
+	if (istype(message, /image) || istype(message, /sound) || istype(target, /savefile) || !(ismob(target) || islist(target) || istype(target, /client) || istype(target, /datum/log) || target == world))
+		target << message
+		if (!istype(target, /atom)) // Really easy to mix these up, and not having to make sure things are mobs makes the code cleaner.
+			CRASH("DEBUG: Boutput called with invalid message")
+		return
+
+	//Otherwise, we're good to throw it at the user
+	else if (istext(message))
+		if (istext(target))
+			return
+
+		//Some macros remain in the string even after parsing and fuck up the eventual output
+		if (findtext(message, "\improper"))
+			message = replacetext(message, "\improper", "")
+		if (findtext(message, "\proper"))
+			message = replacetext(message, "\proper", "")
+
+		//Grab us a client if possible
+		var/client/C
+		if (istype(target, /client))
+			C = target
+		else if (istype(target, /mob))
+			C = target:client
+		else if (istype(target, /datum/mind) && target:current)
+			C = target:current:client
```

Comment:
A single proc to get client/output target on /datum would be faster that all these istypes and :'s ?